### ISSUE TITLE
What features would you like to see?

### ISSUE THREAD

**Issue by szymonmaszke** [szymonmaszke/vimpyter#17]
As the title says, post your ideas below.

**Comment by szymonmaszke** [comment]
## Current ideas in order of importance and amount of additional work needed:

- Major source code refactor
- Binding for inserting above/below/here python block
- Support for [hisaknown/jupyterkernel.vim](https://github.com/hisaknown/jupyterkernel.vim)
- [ultisnips](https://github.com/SirVer/ultisnips) integration (other snippet plugins?)
- Run kernel independently from notebooks (connect to remote kernels as well) (help? xd)
- Closer integration with Jupyter notebooks (maybe using [jupytext](https://github.com/mwouts/jupytext)?)
- Automatic notebook scrolling like (see an example [here](https://github.com/iamcco/markdown-preview.nvim))

**Comment by malramsay64** [comment]
So I had a go at implementing support for jupytext [here](https://github.com/malramsay64/vimpyter). While I did manage to get it working, my opinion is that to properly implement jupytext support requires a rethink of the workflow around using it. Like having the markdown alongside the notebook, that is the the paired representation in jupytext.

An alternative approach is to use a `.py` representation so that all the autocompletion, code linting, code formatting etc, all works without much additional configuration.

**Comment by DerWeh** [comment]
Thanks for your great project.
I think the focus should be to get a more interactive and seamless interaction with the notebooks. Currently it is not really workable. If I edit something in vim, I have to reload my notebook with lots of hassles, then rerun the calculations ... Currently it is still to cumbersome.

`jupyterkernel.vim` seams like one possible direction. I am currently still working with the plain old [vim-ipython](https://github.com/wilywampa/vim-ipython) and simple `.py` files. While it is not nearly as pretty, it simply provides a better workflow (and seamlessly integrates with language-servers).
I think in order to be snappy, the plugin needs some possibility to run a cell from within `vim`. Till then this plugin will just be a way to write some additional text to notebooks.

I am looking forward to your future work, keep going.

**Comment by vnijs** [comment]
You may be interested in "https://github.com/mwouts/jupytext" which makes it possible to "link", e.g., a python or R(markdown) script to a jypyter notebook. You can edit the notebook and the linked file will be auto-updated. You can also edit the linked file, e.g., in (space)vim, and the notebook will be updated as well. Very nice!

I also saw the follow issue posted on 'spacevim' (https://github.com/SpaceVim/SpaceVim/issues/1854) but I don't think this has happened yet but it is listed as a "milestone".

One thing that would be fantastic in vimpyter (and spacevim) would be to send Python (or R) code to a running jupyter kernel. This is something that atom + hydrogen currently does a fantastic job at. See link below

https://nteract.gitbooks.io/hydrogen/docs/Usage/GettingStarted.html

You can even connect to jupyter kernels (e.g., R and Python) running in local or remote docker containers which is something I use all the time. Works really great ... from atom ... but I would *love* to have this work form (space)vim.

https://github.com/nteract/hydrogen/blob/master/docs/Usage/RemoteKernelConnection.md

**Comment by szymonmaszke** [comment]
Yeah, I think removing `notedown` dependency and changing it to [jupytext](https://github.com/mwouts/jupytext) is the road I will take when time allows.

**Comment by wsdjeg** [comment]
@szymonmaszke Thanks for your plugin, I am spacevim author, and I will add new layer for ipynb filetype, and I will fork this plugin, make some change, of cause will sent PR to this repo.

**Comment by anosillus** [comment]
Hello, I like your plugin.
But I wanted to use python lint tools and some python plugins.
So I prefer like to use `ipynb-py-convert` tools instead of notedown.
'ipynb-py-convert' convert an ipynb-file into a python file.(not markdown).

I have folked your plugin and using ipynb-py-convert tools.
If user's want to use this tools. I can send pull requests.

**Comment by tvatter** [comment]
See #21

**Comment by mikh3x4** [comment]
Is work on this plugin still happening? I'm really excited about having vim - jupyter integration and I'm trying to decide on what workflow is best
